concordances/handlers.go

```diff
@@ -0,0 +1,257 @@
+	router.HandleFunc(status.GTGPath, status.NewGoodToGoHandler(healthService.gtgCheck))
+	router.HandleFunc(status.BuildInfoPath, status.BuildInfoHandler)
+
+	monitoringRouter := httphandlers.TransactionAwareRequestLoggingHandler(log.StandardLogger(), router)
+	monitoringRouter = httphandlers.HTTPMetricsHandler(metrics.DefaultRegistry, monitoringRouter)
+	http.Handle("/", monitoringRouter)
+}
+
+func (h *ConcordancesRwHandler) HandleGet(rw http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	//400
+	if h.invalidPath(vars) {
+		h.HandleBadRequest(rw, r)
+		return
+	}
+	uuid := vars[UUID_Param]
+	model, err := h.srv.Read(uuid)
+
+	//503
+	if err != nil {
+		h.handleServiceUnavailable("retrieving", err.Error(), rw, r)
+		return
+	}
+	//404
+	if model.ConcordedIds == nil {
+		h.handleNotFound(uuid, rw, r)
+		return
+	}
+
+	output, err := json.Marshal(&model)
+	//404
+	if err != nil {
+		h.handleNotFound(uuid, rw, r)
+		return
+	}
+	//200
+	rw.Header().Set("Content-Type", ContentTypeJson)
+	rw.WriteHeader(http.StatusOK)
+	rw.Write(output)
+}
+
+func (h *ConcordancesRwHandler) HandlePut(rw http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+
+	//400
+	if h.invalidPath(vars) {
+		h.HandleBadRequest(rw, r)
+		return
+	}
+	uuid := vars[UUID_Param]
+	if r.ContentLength <= 0 {
+		h.handleBadPayload(ErrorMsg_BadJson, rw, r)
+		return
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
+	//400
+	if err != nil {
+		h.handleBadPayload(ErrorMsg_BadJson, rw, r)
+		return
+	}
+
+	model := db.Model{}
+	err = json.Unmarshal(body, &model)
+	//400
+	if err != nil {
+		h.handleBadPayload(ErrorMsg_BadJson, rw, r)
+		return
+	}
+
+	err = h.invalidModel(uuid, model)
+	//400
+	if err != nil {
+		h.handleBadPayload(err.Error(), rw, r)
+		return
+	}
+
+	created, err := h.srv.Write(model)
+
+	//503
+	if err != nil {
+		h.handleServiceUnavailable("storing", err.Error(), rw, r)
+		return
+	}
+
+	if created {
+		rw.WriteHeader(http.StatusCreated)
+		return
+	} else {
+		rw.WriteHeader(http.StatusOK)
+	}
+}
+
+func (h *ConcordancesRwHandler) HandleDelete(rw http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	//400
+	if h.invalidPath(vars) {
+		h.HandleBadRequest(rw, r)
+		return
+	}
+	uuid := vars[UUID_Param]
+	deleted, err := h.srv.Delete(uuid)
+	//503
+	if err != nil {
+		h.handleServiceUnavailable("deleting", err.Error(), rw, r)
+		return
+	}
+	//404
+	if !deleted {
+		h.handleNotFound(uuid, rw, r)
+		return
+	}
+	//204
+	rw.WriteHeader(http.StatusNoContent)
+}
+
+func (h *ConcordancesRwHandler) HandleCount(rw http.ResponseWriter, r *http.Request) {
+	count, err := h.srv.Count()
+	//503
+	if err != nil {
+		h.handleServiceUnavailable("counting", err.Error(), rw, r)
+		return
+	}
+	//200
+	rw.Header().Set("Content-Type", "text/plain")
+	rw.WriteHeader(http.StatusOK)
+	b := []byte{}
+	b = strconv.AppendInt(b, count, 10)
+	rw.Write(b)
+}
+
+func (h *ConcordancesRwHandler) HandleBadRequest(rw http.ResponseWriter, r *http.Request) {
```

Comment:
All these handle methods are virtually the same, can we have just one? In other services it is

func writeJSONError(w http.ResponseWriter, errorMsg string, statusCode int) {
    w.WriteHeader(statusCode)
    fmt.Fprintln(w, fmt.Sprintf("{\"message\": \"%s\"}", errorMsg))
}
Also is the log necessary as the log message should be close to where the error occurs rather than being propagated upwards